tactics/tactics.ml

```diff
@@ -3028,38 +3041,54 @@ let specialize (c,lbind) ipat =
       let typ_of_c_hd = pf_get_type_of gl thd in
-      let lprod, concl = decompose_prod_assum sigma typ_of_c_hd in
+      let (lprod:rel_context), concl = decompose_prod_assum sigma typ_of_c_hd in
       (* lprd = initial products (including letins).
          l(tstack initially) = the same products after unification vs lbind (some metas remain)
          args: accumulator : args to apply to hd: inferred args + metas reabstracted *)
-      let rec rebuild_lambdas sigma lprd args hd l =
+      let rec rebuild sigma concl (lprd:rel_context) (accargs:EConstr.t list)
+                (accprods:rel_context) hd (l:EConstr.t list) =
+        let open Context.Rel.Declaration in
         match lprd , l with
-        | _, [] -> sigma,applist (hd, (List.map (nf_evar sigma) args))
-        | Context.Rel.Declaration.LocalAssum(nme,_)::lp' , t::l' when occur_meta sigma t ->
+        | _, [] -> sigma
+                  , applist (hd, (List.map (nf_evar sigma) (List.rev accargs)))
+                  , EConstr.it_mkProd_or_LetIn concl accprods
+        | (LocalAssum(nme,_) as assum)::lp' , t::l' when occur_meta sigma t ->
           (* nme has not been resolved, let us re-abstract it. Same
              name but type updated by instantiation of other args. *)
           let sigma,new_typ_of_t = Typing.type_of clause.env sigma t in
           let r = Retyping.relevance_of_type env sigma new_typ_of_t in
-          let liftedargs = List.map liftrel args in
           (* lifting rels in the accumulator args *)
-          let sigma,hd' = rebuild_lambdas sigma lp' (liftedargs@[mkRel 1 ]) hd l' in
+          let liftedargs = List.map liftrel accargs in
+          let sigma,hd',prods =
+            rebuild sigma concl lp' (mkRel 1 ::liftedargs) (assum::accprods) hd l' in
           (* replace meta variable by the abstracted variable *)
           let hd'' = subst_term sigma t hd' in
-          (* lambda expansion *)
-          sigma,mkLambda ({nme with binder_relevance=r},new_typ_of_t,hd'')
-        | Context.Rel.Declaration.LocalAssum _::lp' , t::l' ->
-          let sigma,hd' = rebuild_lambdas sigma lp' (args@[t]) hd l' in
-          sigma,hd'
-        | Context.Rel.Declaration.LocalDef _::lp' , _ ->
-           (* letins have been reduced in l and should anyway not
-              correspond to an arg, we ignore them. *)
-          let sigma,hd' = rebuild_lambdas sigma lp' args hd l in
-          sigma,hd'
+          (* we reabstract the non solved argument *)
+          sigma,mkLambda ({nme with binder_relevance=r},new_typ_of_t,hd''),prods
+        | (LocalAssum (nme,tnme))::lp' , t::l' ->
+           (* thie arg was solved, we update thing accordingly *)
+           (* we replace in lprod the arg by rel 1 *)
+           let substlp' = (* rel 1 must be lifted along the context *)
+             map_rel_context_lift (fun i x -> map_constr (replace_term sigma (mkRel i) t) x)
+               env 1 lp' in
```

Comment:
This is my way of replacing a rel by another term (that may reference other rel) in a rel_context.